File: Parcial 3/DPChange.py
import math

c = [1, 3, 7]
d = 3
M = 77
bestNumCoins = []

def DPChange(M, c, d):

    for a in range(M+1):
        bestNumCoins.append(0)
    bestNumCoins[0] = 0

    for m in range(1, M+1, 1):
        bestNumCoins[m] = math.inf
        for i in range(d):
            if m >= c[i]:
                if bestNumCoins[m-c[i]] + 1 < bestNumCoins[m]:
                    bestNumCoins[m] = bestNumCoins[m-c[i]] + 1
    return bestNumCoins[M]

print("Tu cambio es:", DPChange(M, c, d))

File: Parcial 3/DPChange.cpp
#include <iostream>
#define INFINITY 2147483647
using namespace std;

int DPChange(int M, int c[], int d, int bestNumCoins[])
{
    for (int i = 0; i < M; i++)
    {
        bestNumCoins[i] = 0;
    }

    bestNumCoins[0] = 0;

    for (int m = 0; m < M; m++)
    {
        bestNumCoins[m] = INFINITY;
        for (int i = 0; i < d; i++)
        {
            if (m >= c[i])
            {
                if (bestNumCoins[m - c[i]] + 1 < bestNumCoins[m])
                {
                    bestNumCoins[m] = bestNumCoins[m - c[i]] + 1;
                }
            }
        }
    }

    return *bestNumCoins;
}

int main()
{
    int c[3] = {1, 3, 7};
    int d = 3;
    int M = 3;
    int bestNumCoins[M];

    DPChange(M, c, d, bestNumCoins);

    cout << "BetterChange = [ ";

    for (int j = 0; j < M; j++)
    {
        cout << bestNumCoins[j] << ", ";
    }
    cout << "]" << endl;

    return 0;
}